exchanges/orderbook/incremental_updates.go

```diff
@@ -0,0 +1,234 @@
+	errRESTSnapshot           = errors.New("cannot update REST protocol loaded snapshot")
+	errChecksumMismatch       = errors.New("checksum mismatch")
+	errChecksumGeneratorUnset = errors.New("checksum generator unset")
+)
+
+// Update holds changes that are to be applied to a stored orderbook
+type Update struct {
+	UpdateID       int64
+	UpdateTime     time.Time
+	UpdatePushedAt time.Time
+	Asset          asset.Item
+	Bids           []Tranche
+	Asks           []Tranche
+	Pair           currency.Pair
+
+	// ExpectedChecksum defines the expected value when the books have been verified
+	ExpectedChecksum uint32
+	// GenerateChecksum is a function that will be called to generate a checksum from the stored orderbook post update
+	GenerateChecksum func(snapshot *Base) uint32
+	// AllowEmpty, when true, permits loading an empty order book update to set an UpdateID without including actual data
+	AllowEmpty bool
+	// Action defines the action to be performed on the orderbook e.g. amend, delete, insert, update/insert
+	// Orderbook IDs are used to identify the orderbook level to be updated, deleted or inserted
+	Action Action
+
+	SkipOutOfOrderLastUpdateID bool
+}
+
+// ProcessUpdate applies updates to the orderbook depth, on error it will invalidate the orderbook and return the
+// error, this is to ensure the orderbook is always in a valid state.
+func (d *Depth) ProcessUpdate(u *Update) error {
+	if len(u.Bids) == 0 && len(u.Asks) == 0 && !u.AllowEmpty {
+		return d.Invalidate(ErrEmptyUpdate)
+	}
+
+	// TODO: Enforce UpdatePushedAt set to determine server latency
+
+	d.m.Lock()
+	defer d.m.Unlock()
+
+	if d.validationError != nil {
+		return d.validationError
+	}
+
+	// This will process out of order updates but will not error on them.
+	// TODO: Error on out of order updates; this is intentionally kept as is from the buffer package.
+	// Add update.UpdateTime time check to ensure that the update is newer than the last update,
+	// this should screen zero values as well.
+	if u.SkipOutOfOrderLastUpdateID && d.lastUpdateID >= u.UpdateID {
+		return nil
+	}
+
+	if d.options.restSnapshot {
+		return d.invalidate(errRESTSnapshot)
+	}
+
+	if u.Action != 0 {
+		if err := d.updateByIDAndAction(u); err != nil {
+			return d.invalidate(err)
+		}
+	} else {
+		if err := d.updateBidAskByPrice(u); err != nil {
+			return d.invalidate(err)
+		}
+	}
+
+	if u.ExpectedChecksum != 0 {
+		if u.GenerateChecksum == nil {
+			return d.invalidate(errChecksumGeneratorUnset)
+		}
+		if checksum := u.GenerateChecksum(d.snapshot()); checksum != u.ExpectedChecksum {
+			return d.invalidate(fmt.Errorf("%s %s %s %w: expected '%d', got '%d'", d.exchange, d.pair, d.asset, errChecksumMismatch, u.ExpectedChecksum, checksum))
+		}
+	} else if d.verifyOrderbook {
+		if err := verify(d.snapshot()); err != nil {
+			return d.invalidate(err)
+		}
+	}
+
+	return nil
+}
+
+// // TODO: Confirm no alloc as it shouldn't escape
```

Comment:
🚧 What does this mean?
🦍 Double `//`

Reply:
oooh! Something I didn't do, wanted to check if this struct was actually allocated or was inlined, I will check it soon and RM

Reply:
🚧 Still outstanding, right?

Reply:
Yeah 😭 soz

Reply:
https://github.com/thrasher-corp/gocryptotrader/pull/1908/commits/24cdb863ff59e21fdc6c0867b12ab2ea0c0fe32e